From e56132d4fef2bb66d6c2c4d52b2684dba13577df Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@foss.st.com>
Date: Fri, 7 Jun 2024 17:37:40 +0200
Subject: [PATCH 619/841] feat(stm32mp1): make stpmic2 usable for STM32MP1

Update STPMIC2 driver to use it on STM32MP1 especially
STM32MP_STPMIC1L.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 1d95794a05e6124bb894ed612cd8176cc7827355 feat(stm32mp1): make stpmic2 usable for STM32MP1]
Change-Id: I0db727a093a6a85dca7a74be280c0d1af0e54417
Signed-off-by: Pascal Paillet <p.paillet@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/386158
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Pascal PAILLET <p.paillet@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Tested-by: Pascal PAILLET <p.paillet@st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
---
 docs/plat/st/stm32mp1.rst       |  2 ++
 drivers/st/pmic/stm32mp_pmic2.c | 21 +++++++++++++++++++++
 drivers/st/pmic/stpmic2.c       |  5 +++++
 include/drivers/st/stpmic2.h    |  4 ++++
 plat/st/stm32mp1/platform.mk    | 14 ++++++++++++--
 5 files changed, 44 insertions(+), 2 deletions(-)

diff --git a/docs/plat/st/stm32mp1.rst b/docs/plat/st/stm32mp1.rst
index 40b498748..8f0921ec1 100644
--- a/docs/plat/st/stm32mp1.rst
+++ b/docs/plat/st/stm32mp1.rst
@@ -100,6 +100,8 @@ Dedicated STM32MP1 flags:
   | Default: 0
 - | ``STM32MP15``: to select STM32MP15 variant configuration.
   | Default: 1
+- | ``STM32MP_STPMIC1L``: to select STPMIC1L variant configuration.
+  | Default: 0
 
 
 Boot with FIP
diff --git a/drivers/st/pmic/stm32mp_pmic2.c b/drivers/st/pmic/stm32mp_pmic2.c
index 27ac51206..247b342ae 100644
--- a/drivers/st/pmic/stm32mp_pmic2.c
+++ b/drivers/st/pmic/stm32mp_pmic2.c
@@ -181,6 +181,27 @@ bool initialize_pmic_i2c(void)
 	return true;
 }
 
+void pmic_switch_off(void)
+{
+	if (stpmic2_switch_off(pmic2) == 0) {
+		udelay(100);
+	}
+
+	/* Shouldn't be reached */
+	panic();
+}
+
+#if DEBUG
+void print_pmic_info_and_debug(void)
+{
+}
+#endif
+
+int pmic_voltages_init(void)
+{
+	return 0;
+}
+
 static int pmic2_set_state(const struct regul_description *desc, bool enable)
 {
 	struct regul_handle_s *regul = (struct regul_handle_s *)desc->driver_data;
diff --git a/drivers/st/pmic/stpmic2.c b/drivers/st/pmic/stpmic2.c
index cd0eca97a..9bb4efa18 100644
--- a/drivers/st/pmic/stpmic2.c
+++ b/drivers/st/pmic/stpmic2.c
@@ -456,6 +456,11 @@ int stpmic2_regulator_set_prop(struct pmic_handle_s *pmic, uint8_t id,
 	return -EPERM;
 }
 
+int stpmic2_switch_off(struct pmic_handle_s *pmic)
+{
+	return stpmic2_register_update(pmic, MAIN_CR, SWOFF, SWOFF);
+}
+
 #if EVENT_LOG_LEVEL == LOG_LEVEL_VERBOSE
 void stpmic2_dump_regulators(struct pmic_handle_s *pmic)
 {
diff --git a/include/drivers/st/stpmic2.h b/include/drivers/st/stpmic2.h
index c84f7703a..02af038f7 100644
--- a/include/drivers/st/stpmic2.h
+++ b/include/drivers/st/stpmic2.h
@@ -182,6 +182,9 @@ enum {
 #define PMIC_REF_ID_STPMIC25	U(2)
 #define PMIC_REF_ID_STPMIC2L	U(3)
 
+/* MAIN_CR bits definition */
+#define SWOFF			BIT(0)
+
 /* BUCKS_MRST_CR bits definition */
 #define BUCK1_MRST		BIT(0)
 #define BUCK2_MRST		BIT(1)
@@ -352,4 +355,5 @@ int stpmic2_regulator_set_prop(struct pmic_handle_s *pmic, uint8_t id,
 			       enum stpmic2_prop_id prop, uint32_t arg);
 
 int stpmic2_is_buck1_high_voltage(struct pmic_handle_s *pmic, bool *high);
+int stpmic2_switch_off(struct pmic_handle_s *pmic);
 #endif /*STPMIC2_H*/
diff --git a/plat/st/stm32mp1/platform.mk b/plat/st/stm32mp1/platform.mk
index ae0c711fe..ff4188f7c 100644
--- a/plat/st/stm32mp1/platform.mk
+++ b/plat/st/stm32mp1/platform.mk
@@ -37,6 +37,8 @@ STM32MP13		:=	0
 STM32MP15		:=	1
 endif
 
+STM32MP_STPMIC1L	?=	0
+
 ifeq ($(STM32MP13),1)
 # Will use SRAM2 as mbedtls heap
 STM32MP_USE_EXTERNAL_HEAP :=	1
@@ -169,6 +171,7 @@ $(eval $(call assert_booleans,\
 		STM32MP_DDR_32BIT_INTERFACE \
 		STM32MP_DDR_DUAL_AXI_PORT \
 		STM32MP_SSP \
+		STM32MP_STPMIC1L \
 		STM32MP_USE_EXTERNAL_HEAP \
 		STM32MP13 \
 		STM32MP15 \
@@ -202,6 +205,7 @@ $(eval $(call add_defines,\
 		STM32MP_DDR_32BIT_INTERFACE \
 		STM32MP_DDR_DUAL_AXI_PORT \
 		STM32MP_SSP \
+		STM32MP_STPMIC1L \
 		STM32MP_USE_EXTERNAL_HEAP \
 		STM32MP13 \
 		STM32MP15 \
@@ -226,14 +230,20 @@ PLAT_BL_COMMON_SOURCES	+=	drivers/arm/tzc/tzc400.c				\
 				drivers/st/ddr/stm32mp1_ddr_helpers.c			\
 				drivers/st/i2c/stm32_i2c.c				\
 				drivers/st/iwdg/stm32_iwdg.c				\
-				drivers/st/pmic/stm32mp_pmic.c				\
-				drivers/st/pmic/stpmic1.c				\
 				drivers/st/reset/stm32mp1_reset.c			\
 				plat/st/stm32mp1/stm32mp1_context.c			\
 				plat/st/stm32mp1/stm32mp1_dbgmcu.c			\
 				plat/st/stm32mp1/stm32mp1_helper.S			\
 				plat/st/stm32mp1/stm32mp1_syscfg.c
 
+ifeq ($(STM32MP_STPMIC1L),1)
+PLAT_BL_COMMON_SOURCES	+=	drivers/st/pmic/stm32mp_pmic2.c				\
+				drivers/st/pmic/stpmic2.c
+else
+PLAT_BL_COMMON_SOURCES	+=	drivers/st/pmic/stm32mp_pmic.c				\
+				drivers/st/pmic/stpmic1.c
+endif
+
 PLAT_BL_COMMON_SOURCES  +=	drivers/st/nvmem/stm32mp_tamp_nvram_mp1.c
 
 ifeq ($(STM32MP13),1)
-- 
2.47.2

